src/rules/selector-max-depth/index.js

```diff
@@ -0,0 +1,78 @@
+import resolvedNestedSelector from "postcss-resolve-nested-selector"
+
+import {
+  isStandardRule,
+  isStandardSelector,
+  report,
+  ruleMessages,
+  validateOptions,
+} from "../../utils"
+
+export const ruleName = "selector-max-depth"
+
+export const messages = ruleMessages(ruleName, {
+  expected: (selector, depth) => `Expected "${selector}" to have a depth equal to or less than ${depth}`,
+})
+
+export default function (max) {
+  return (root, result) => {
+    const innerSelectorReg = /(?::not\((.*?)\))/g
+    const validOptions = validateOptions(result, ruleName, {
+      actual: max,
+      possible: [function (max) {
+        return typeof max === "number" && max > 0
+      }],
+    })
+    if (!validOptions) { return }
+
+    // Compares depth of a selector with the maximum
+    function checkDepth(selector, rule) {
```

Comment:
I don't think the docs are misleading at all. They do not say "Do anything necessary to get the fastest possible performance, at the expense of anything else." They just say "Try out ways to speed up rules, and use benchmarking to test if they're working." Given some stats, it's a judgment call to determine how the final code should look.